### PR TITLE
Enh/wait

### LINKED ISSUE #3014: Error checking MapNode hash with fMRIPrep
### Summary

```
190904-01:20:18,695 nipype.workflow ERROR:
	 Node t1_conform failed to run on host 890949213e2e.
190904-01:20:18,696 nipype.workflow ERROR:
	 Saving crash info to /tmp/ds054/derivatives/smriprep/sub-100185/log/20190904-011955_7143d03c-6f71-442e-b03b-868465d5d7f5/crash-20190904-012018-UID1001-t1_conform-79d9866d-bc88-4848-9512-c54ff701db1b.txt
Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/multiproc.py", line 269, in _send_procs_to_workers
    num_subnodes = self.procs[jobid].num_subnodes()
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 1204, in num_subnodes
    self._get_inputs()
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 1219, in _get_inputs
    super(MapNode, self)._get_inputs()
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 521, in _get_inputs
    outputs = _load_resultfile(results_file)[0].outputs
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/utils.py", line 311, in load_resultfile
    if resolve and result.outputs:
AttributeError: 'NoneType' object has no attribute 'outputs'
```
https://circleci.com/gh/oesteban/smriprep/502

Related to #3009 (in that case, it occurred for a regular Node).

### LINKED ISSUE #3076: Strange results file checking behavior that breaks workflows at nested mapnodes in nipype 1.2.3
### Summary
Nested MapNodes hang in some workflows when trying to load the node's result file. It appears this is because results.outputs.get() fails since there is a Bunch representation of outputs, rather than a dictionary representation.

### Manifests as follows:
At the logging debug level during execution:
```
Outputs object of loaded result /private/tmp/work/20191012-164048_df69a011-5615-4010-aa22-d4a0c94c5696_wf_single_subject_fmri_0025427/wf_single_sub_0025427_fmri_20191012-164047/meta_wf_0025427/fmri_connectometry_0025427/thresh_func_node/result_thresh_func_node.pklz is a Bunch.
```

and then  if I attempt to load the file manually:
```
In [32]: res.outputs.get()                                                                                                                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-347df97b2ed0> in <module>
----> 1 res.outputs.get()

/usr/local/anaconda3/lib/python3.7/site-packages/nipype-2.0.0.dev0+g7b9c63932-py3.7.egg/nipype/interfaces/base/support.py in get(self, *args)

TypeError: get expected at least 1 arguments, got 0
```
See attached example results file:
[result_thresh_func_node.pklz.zip](https://github.com/nipy/nipype/files/3721280/result_thresh_func_node.pklz.zip)

Interestingly, it seems that `load_resultfile` from `nipyp/epipeline/engine/utils.py` fails, but only for the result file of the primary node (i.e. and not its subnodes if they get parameterized)

Appears to only occurs as of version nipype 1.2.3. Looks like this might also be related to #3009 #3014 #3075 and/or #1798 of fmriprep.

Any ideas @oesteban ?

@dPys

### LINKED ISSUE #3081: No such file or directory: 'result_bidssrc.pklz'
Hello everyone,

### Summary
I am running fmriprep on a remote server.

### Actual behavior
During running, I have several time the same error:

"
Error while checking node hash, forcing re-run. Although this error may not prevent the workflow from running, it could indicate a major problem. Please report a new issue at https://github.com/nipy/nipype/issues adding the following information:

Node: fmriprep_wf.single_subject_p001_wf.bids_info
	Interface: niworkflows.interfaces.bids.BIDSInfo
	Traceback:
Traceback (most recent call last):

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/base.py", line 337, in _local_hash_check
    cached, updated = self.procs[jobid].is_cached()

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 297, in is_cached
    hashed_inputs, hashvalue = self._get_hashval()

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 494, in _get_hashval
    self._get_inputs()

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 521, in _get_inputs
    outputs = _load_resultfile(results_file)[0].outputs

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/utils.py", line 300, in load_resultfile
    result = loadpkl(results_file)

  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/utils/filemanip.py", line 687, in loadpkl
    pkl_file = pklopen(infile.name, 'rb')

  File "/usr/local/miniconda/lib/python3.7/gzip.py", line 53, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)

  File "/usr/local/miniconda/lib/python3.7/gzip.py", line 163, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')

FileNotFoundError: [Errno 2] No such file or directory: 'result_bidssrc.pklz'
"

### Script/Workflow details
The main command is: docker run --rm -e DOCKER_VERSION_8395080871=19.03.2 -v /opt/freesurfer/license.txt:/opt/freesurfer/license.txt:ro -v /path_to_data:/data:ro -v /path_to_data_output:/out poldracklab/fmriprep:1.5.0 /data /out participant --participant_label sub-p001 --fs-no-reconall

Thank you very much in advance!

### PR DESCRIPTION
**Pull request by satra**
## Summary
Continuation of #3083 

this adds an explicit timeout in reading the results file that may be missing due to file system delays.

Closes #3076

Hopefully:

Fixes #3014
Fixes #3081

## List of changes proposed in this PR (pull-request)
- Add waiting for result in reading pickled files.

## Acknowledgment

- [X] \(Mandatory\) I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by satra** [comment]
@oesteban @dpys - could you try this branch? this only handles any results file issue, not resource management.

File: nipype/utils/filemanip.py [hunk 1]
```diff
@@ -690,32 +701,50 @@ def load_json(filename):
...
+    t = time()
+    timeout = float(config.get("execution", "job_finished_timeout"))
+    timed_out = True
+    while (time() - t) < timeout:
+        try:
+            glob(str(infile)).pop()
+            timed_out = False
+            break
+        except Exception as e:
+            fmlogger.debug(e)
```
Comment (effigies) [inline nipype/utils/filemanip.py]: Is there something special about `glob`, as opposed to, e.g.,

```Python
if infile.exists():
    timed_out = False
    break
fmlogger.debug("'{}' missing; waiting 2s".format(infile))
```

This feels more straightforward, so if `glob` is doing something useful beyond an `exists()` check, a comment would be useful.

File: nipype/utils/filemanip.py [hunk 2]
```diff
@@ -690,32 +701,50 @@ def load_json(filename):
...
+        sleep(2)
+    if timed_out:
+        error_message = (
+            "Ressult file {0} expected, but "
```
Comment (effigies) [inline nipype/utils/filemanip.py]:
```suggestion
            "Result file {0} expected, but "
```

File: nipype/utils/filemanip.py (hunk 1 above)
Comment (satra) [reply nipype/utils/filemanip.py]: updated the code - that code is old, and i don't remember why we used glob.

**Comment by dPys** [comment]
Thanks @satra -- this patch appears to fix #3076!

File: nipype/utils/filemanip.py [hunk 3]
```diff
@@ -779,50 +813,53 @@ def read_stream(stream, logger=None, encoding=None):
...
+
+            metadata = json.dumps({"version": version})
 
-                pkl_file.write(metadata.encode('utf-8'))
-                pkl_file.write('\n'.encode('utf-8'))
+            pkl_file.write(metadata.encode("utf-8"))
+            pkl_file.write("\n".encode("utf-8"))
 
-            pickle.dump(record, pkl_file)
+        pickle.dump(record, pkl_file)
```
Comment (oesteban) [inline nipype/utils/filemanip.py]: Could we do this to a bytes IO stream outside the context and then open the context and just write it?

Comment (satra) [reply nipype/utils/filemanip.py]: it is a bytes io stream and is being done with the context.

Comment (satra) [reply nipype/utils/filemanip.py]: see 827-828

Comment (oesteban) [reply nipype/utils/filemanip.py]: I meant something like this:

```Python
from io import BytesIO

    with BytesIO() as f:
        if versioning:
            from nipype import __version__ as version
            metadata = json.dumps({"version": version})
            f.write(metadata.encode("utf-8"))
            f.write("\n".encode("utf-8"))
        pickle.dump(record, f)
        content = f.get_value()

    pkl_open = gzip.open if filename.endswith(".pklz") else open
    with pkl_open(filename, "wb") as pkl_file:
        pkl_file.write(content)
```

Comment (effigies) [reply nipype/utils/filemanip.py]: Why?

Comment (oesteban) [reply nipype/utils/filemanip.py]: minimize being inside the context with an open file?

Comment (effigies) [reply nipype/utils/filemanip.py]: Maybe I'm missing something. Are we worried about some kind of race here? If so, why drop the lock?

Comment (oesteban) [reply nipype/utils/filemanip.py]: Not really. Separating steps just seemed a good idea to me in this case. We haven't tested versioning thoroughly, and doing the pickle.dump to memory seemed more efficient than all at once.

Also if we keep seeing issues, separating both will make it easier to make more targeted interventions (let's imagine we discover that gzip.open does not get along with NFS)

Comment (oesteban) [reply nipype/utils/filemanip.py]: (that said, I agree this is not strictly necessary for this PR, and at the risk of introducing some other bug I am happy dismissing).

**Comment by satra** [comment]
i refactored `savepkl`. and yes there are more stylistic changes, since i have pre-commit turned on.

**Comment by satra** [comment]
here is the direct pointer to the function: https://github.com/nipy/nipype/pull/3089/files#diff-7d84fc0957a520f25bb02479beb87c51R825

**Comment by satra** [comment]
@effigies @oesteban - can we merge this and cut a new release/release candidate?

**Comment by effigies** [comment]
No objections.

**Comment by satra** [comment]
@oesteban - can i merge this?

**Comment by oesteban** [comment]
Sorry for holding this up

**Pull request merged**